File: sdks/typescript/src/sse.ts
import { BridgeAPIError, BridgeProtocolError, record, string } from "./errors.js";
import type { ImageResponse, StreamEvent } from "./types.js";

export class SseDecoder {
  readonly #decoder = new TextDecoder("utf-8", { fatal: true });
  readonly #maximumLineBytes: number;
  readonly #maximumEventBytes: number;
  #lineParts: Uint8Array[] = [];
  #lineBytes = 0;
  #pendingCr = false;
  #eventName = "message";
  #data: string[] = [];
  #size = 0;

  constructor(maximumLineBytes: number, maximumEventBytes: number) {
    for (const [value, name] of [
      [maximumLineBytes, "maximumLineBytes"],
      [maximumEventBytes, "maximumEventBytes"],
    ] as const) {
      if (!Number.isSafeInteger(value) || value <= 0) {
        throw new RangeError(`${name} must be a positive safe integer`);
      }
    }
    this.#maximumLineBytes = maximumLineBytes;
    this.#maximumEventBytes = maximumEventBytes;
  }

  push(chunk: Uint8Array): StreamEvent[] {
    const events: StreamEvent[] = [];
    let start = 0;
    for (let index = 0; index < chunk.byteLength; index += 1) {
      const byte = chunk[index];
      if (this.#pendingCr) {
        const event = this.#processLine();
        if (event) events.push(event);
        this.#pendingCr = false;
        if (byte === 0x0a) {
          start = index + 1;
          continue;
        }
        start = index;
      }
      if (byte === 0x0d || byte === 0x0a) {
        this.#appendLineSegment(chunk.subarray(start, index));
        if (byte === 0x0d) this.#pendingCr = true;
        else {
          const event = this.#processLine();
          if (event) events.push(event);
        }
        start = index + 1;
      }
    }
    this.#appendLineSegment(chunk.subarray(start));
    return events;
  }

  finish(): StreamEvent[] {
    const events: StreamEvent[] = [];
    if (this.#pendingCr || this.#lineBytes > 0) {
      const event = this.#processLine();
      if (event) events.push(event);
      this.#pendingCr = false;
    }
    if (this.#data.length > 0) events.push(this.#dispatch());
    return events;
  }

  #appendLineSegment(segment: Uint8Array): void {
    if (segment.byteLength > this.#maximumLineBytes - this.#lineBytes) {
      throw new BridgeProtocolError("bridge SSE line exceeded the configured byte limit");
    }
    if (segment.byteLength > 0) this.#lineParts.push(segment.slice());
    this.#lineBytes += segment.byteLength;
  }

  #processLine(): StreamEvent | null {
    const bytes = new Uint8Array(this.#lineBytes);
    let offset = 0;
    for (const part of this.#lineParts) {
      bytes.set(part, offset);
      offset += part.byteLength;
    }
    const lineBytes = this.#lineBytes;
    this.#lineParts = [];
    this.#lineBytes = 0;
    let line: string;
    try {
      line = this.#decoder.decode(bytes);
    } catch (error) {
      throw new BridgeProtocolError("bridge returned invalid UTF-8 in the SSE stream", {
        cause: error,
      });
    }
    if (line === "") return this.#data.length > 0 ? this.#dispatch() : null;
    if (line.startsWith(":")) return null;
    this.#size += lineBytes;
    if (this.#size > this.#maximumEventBytes) {
      throw new BridgeProtocolError("bridge SSE event exceeded the configured byte limit");
    }
    const separator = line.indexOf(":");
    const field = separator < 0 ? line : line.slice(0, separator);
    let value = separator < 0 ? "" : line.slice(separator + 1);
    if (value.startsWith(" ")) value = value.slice(1);
    if (field === "event") this.#eventName = value;
    if (field === "data") this.#data.push(value);
    return null;
  }

  #dispatch(): StreamEvent {
    const eventName = this.#eventName;
    const data = this.#data.join("\n");
    this.#eventName = "message";
    this.#data = [];
    this.#size = 0;
    let payload: unknown;
    try {
      payload = JSON.parse(data);
    } catch (error) {
      throw new BridgeProtocolError("bridge returned invalid JSON in an SSE event", {
        cause: error,
      });
    }
    if (eventName === "error") throw BridgeAPIError.fromPayload(200, payload);
    const value = record(payload);
    const type = string(value?.type);
    if (type === "started") return { type };
    if (type === "progress" && string(value?.stage))
      return { type, stage: string(value?.stage) ?? "" };
    if (
      type === "partial_image" &&
      typeof value?.index === "number" &&
      typeof value.partial_index === "number" &&
      typeof value.b64_json === "string"
    ) {
      return {
        type,
        index: value.index,
        partial_index: value.partial_index,
        b64_json: value.b64_json,
      };
    }
    const response = record(value?.response);
    if (type === "completed" && response) {
      return { type, response: response as unknown as ImageResponse };
    }
    throw new BridgeProtocolError(`bridge returned unsupported SSE event type ${String(type)}`);
  }
}

File: sdks/python/src/imagegen_bridge/_sse.py
"""Bounded SSE decoders for sync and async HTTPX line iterators."""

from __future__ import annotations

import json
from collections.abc import AsyncIterable, AsyncIterator, Iterable, Iterator
from typing import Any

from ._errors import BridgeAPIError, BridgeProtocolError
from ._types import (
    CompletedEvent,
    ImageResponse,
    PartialImageEvent,
    ProgressEvent,
    StartedEvent,
    StreamEvent,
)


def _decode(event_name: str, data: str) -> StreamEvent:
    try:
        value: Any = json.loads(data)
    except (json.JSONDecodeError, UnicodeError) as error:
        raise BridgeProtocolError("bridge returned invalid JSON in an SSE event") from error
    if event_name == "error":
        raise BridgeAPIError.from_payload(200, value)
    if not isinstance(value, dict):
        raise BridgeProtocolError("bridge returned a non-object SSE event")
    event_type = value.get("type")
    if event_type == "started":
        return StartedEvent()
    if event_type == "progress" and isinstance(value.get("stage"), str):
        return ProgressEvent(stage=value["stage"])
    if event_type == "partial_image":
        try:
            return PartialImageEvent(
                index=int(value["index"]),
                partial_index=int(value["partial_index"]),
                b64_json=str(value["b64_json"]),
            )
        except (KeyError, TypeError, ValueError) as error:
            raise BridgeProtocolError("bridge returned an invalid partial-image event") from error
    if event_type == "completed" and isinstance(value.get("response"), dict):
        try:
            return CompletedEvent(ImageResponse.from_dict(value["response"]))
        except (KeyError, TypeError, ValueError) as error:
            raise BridgeProtocolError("bridge returned an invalid completion event") from error
    raise BridgeProtocolError(f"bridge returned unsupported SSE event type {event_type!r}")


def _line_text(line: bytearray) -> str:
    try:
        return line.decode("utf-8")
    except UnicodeDecodeError as error:
        raise BridgeProtocolError("bridge returned invalid UTF-8 in the SSE stream") from error


def _iter_sse_lines(chunks: Iterable[bytes], maximum_line_bytes: int) -> Iterator[str]:
    line = bytearray()
    pending_cr = False
    for chunk in chunks:
        for byte in chunk:
            if pending_cr:
                yield _line_text(line)
                line.clear()
                pending_cr = False
                if byte == 0x0A:
                    continue
            if byte == 0x0D:
                pending_cr = True
            elif byte == 0x0A:
                yield _line_text(line)
                line.clear()
            else:
                if len(line) >= maximum_line_bytes:
                    raise BridgeProtocolError("bridge SSE line exceeded the configured byte limit")
                line.append(byte)
    if pending_cr:
        yield _line_text(line)
        line.clear()
    if line:
        yield _line_text(line)


async def _aiter_sse_lines(
    chunks: AsyncIterable[bytes], maximum_line_bytes: int
) -> AsyncIterator[str]:
    line = bytearray()
    pending_cr = False
    async for chunk in chunks:
        for byte in chunk:
            if pending_cr:
                yield _line_text(line)
                line.clear()
                pending_cr = False
                if byte == 0x0A:
                    continue
            if byte == 0x0D:
                pending_cr = True
            elif byte == 0x0A:
                yield _line_text(line)
                line.clear()
            else:
                if len(line) >= maximum_line_bytes:
                    raise BridgeProtocolError("bridge SSE line exceeded the configured byte limit")
                line.append(byte)
    if pending_cr:
        yield _line_text(line)
        line.clear()
    if line:
        yield _line_text(line)


def iter_sse(
    chunks: Iterable[bytes], maximum_line_bytes: int, maximum_event_bytes: int
) -> Iterator[StreamEvent]:
    event_name = "message"
    data: list[str] = []
    size = 0
    for line in _iter_sse_lines(chunks, maximum_line_bytes):
        if line == "":
            if data:
                yield _decode(event_name, "\n".join(data))
            event_name, data, size = "message", [], 0
            continue
        if line.startswith(":"):
            continue
        size += len(line.encode("utf-8"))
        if size > maximum_event_bytes:
            raise BridgeProtocolError("bridge SSE event exceeded the configured byte limit")
        field, separator, value = line.partition(":")
        if separator and value.startswith(" "):
            value = value[1:]
        if field == "event":
            event_name = value
        elif field == "data":
            data.append(value)
    if data:
        yield _decode(event_name, "\n".join(data))


async def aiter_sse(
    chunks: AsyncIterable[bytes], maximum_line_bytes: int, maximum_event_bytes: int
) -> AsyncIterator[StreamEvent]:
    event_name = "message"
    data: list[str] = []
    size = 0
    async for line in _aiter_sse_lines(chunks, maximum_line_bytes):
        if line == "":
            if data:
                yield _decode(event_name, "\n".join(data))
            event_name, data, size = "message", [], 0
            continue
        if line.startswith(":"):
            continue
        size += len(line.encode("utf-8"))
        if size > maximum_event_bytes:
            raise BridgeProtocolError("bridge SSE event exceeded the configured byte limit")
        field, separator, value = line.partition(":")
        if separator and value.startswith(" "):
            value = value[1:]
        if field == "event":
            event_name = value
        elif field == "data":
            data.append(value)
    if data:
        yield _decode(event_name, "\n".join(data))
